server/src/main/java/org/opensearch/index/store/remote/directory/RemoteSnapshotDirectoryFactory.java

```diff
@@ -0,0 +1,88 @@
+ *
+ * @opensearch.internal
+ */
+public final class RemoteSnapshotDirectoryFactory implements IndexStorePlugin.DirectoryFactory {
+    public static final String LOCAL_STORE_LOCATION = "RemoteLocalStore";
+
+    private final Supplier<RepositoriesService> repositoriesService;
+    private final ThreadPool threadPool;
+
+    public RemoteSnapshotDirectoryFactory(Supplier<RepositoriesService> repositoriesService, ThreadPool threadPool) {
+        this.repositoriesService = repositoriesService;
+        this.threadPool = threadPool;
+    }
+
+    @Override
+    public Directory newDirectory(IndexSettings indexSettings, ShardPath localShardPath) throws IOException {
+        final String repositoryName = IndexSettings.SEARCHABLE_SNAPSHOT_REPOSITORY.get(indexSettings.getSettings());
+        final Repository repository = repositoriesService.get().repository(repositoryName);
+        assert repository instanceof BlobStoreRepository : "repository should be instance of BlobStoreRepository";
+        final BlobStoreRepository blobStoreRepository = (BlobStoreRepository) repository;
+        try {
+            return createRemoteSnapshotDirectoryFromSnapshot(indexSettings, localShardPath, blobStoreRepository).get();
+        } catch (InterruptedException | ExecutionException e) {
+            throw new IllegalStateException(e);
+        }
+    }
+
+    private Future<RemoteSnapshotDirectory> createRemoteSnapshotDirectoryFromSnapshot(
+        IndexSettings indexSettings,
+        ShardPath localShardPath,
+        BlobStoreRepository blobStoreRepository
+    ) throws IOException {
+        final BlobPath blobPath = new BlobPath().add("indices")
+            .add(IndexSettings.SEARCHABLE_SNAPSHOT_INDEX_ID.get(indexSettings.getSettings()))
+            .add(Integer.toString(localShardPath.getShardId().getId()));
+        final SnapshotId snapshotId = new SnapshotId(
+            IndexSettings.SEARCHABLE_SNAPSHOT_ID_NAME.get(indexSettings.getSettings()),
+            IndexSettings.SEARCHABLE_SNAPSHOT_ID_UUID.get(indexSettings.getSettings())
+        );
+        Path localStorePath = localShardPath.getDataPath().resolve(LOCAL_STORE_LOCATION);
+        FSDirectory localStoreDir = FSDirectory.open(Files.createDirectories(localStorePath));
+        // make sure directory is flushed to persistent storage
+        localStoreDir.syncMetaData();
+        // this trick is needed to bypass assertions in BlobStoreRepository::assertAllowableThreadPools in case of node restart and a remote
+        // index restore is invoked
+        return threadPool.executor(ThreadPool.Names.SNAPSHOT).submit(() -> {
+            final BlobContainer blobContainer = blobStoreRepository.blobStore().blobContainer(blobPath);
+            final BlobStoreIndexShardSnapshot snapshot = blobStoreRepository.loadShardSnapshot(blobContainer, snapshotId);
+            TransferManager transferManager = new TransferManager(blobContainer, threadPool.executor(ThreadPool.Names.SEARCH));
```

Comment:
Aren't we performing remote I/O on `SEARCH ` threadpool? Can we create a dedicated threadpool for all I/O interactions

Reply:
we agreed that at this phase, we will be using existing thread pools and not create new one until we have a benchmark which will drive this decision. So for now yes we will be using search thread pool except when creating the blob container to  bypass `BlobStoreRepository::assertSnapshotOrGenericThread` checks without adding search threadpool to the assertion